libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp

```diff
@@ -269,15 +256,58 @@ namespace meshkernel
         /// @return a vector of M nodes
         std::vector<Point> GetNodeVectorAtN(UInt n) const { return lin_alg::MatrixColToSTLVector(m_gridNodes, n); }
 
-    private:
-        /// @brief Remove invalid nodes.
-        /// This function is recursive
-        /// @param[in] invalidNodesToRemove Whether there are still invalid nodes to remove
-        void RemoveInvalidNodes(bool invalidNodesToRemove);
+        /// @brief Get the projection
+        /// @return The curvilinear grid projection
+        [[nodiscard]] const Projection& projection() const { return m_projection; }
 
-        /// @brief Computes the valid grid faces
-        void ComputeGridFacesMask();
+        /// @brief Computes the number of nodes
+        [[nodiscard]] UInt GetNumNodes() const { return NumN() * NumM(); }
+
+        /// @brief Computes the number of edges
+        [[nodiscard]] UInt GetNumEdges() const { return NumM() * (NumN() - 1) +
+                                                        (NumM() - 1) * NumN(); }
+
+        /// @brief Computes the number of faces
+        [[nodiscard]] UInt GetNumFaces() const { return (NumN() - 1) * (NumM() - 1); }
+
+        /// @brief Computes the node from an index
+        [[nodiscard]] const Point& Node(const UInt index) const
+        {
+            const auto nodeIndex = GetNodeIndex(index);
+            return m_gridNodes(nodeIndex.m_n, nodeIndex.m_m);
+        }
+
+        /// @brief Get the node CurvilinearGridNodeIndices from a position
+        /// @return The CurvilinearGridNodeIndices
+        CurvilinearGridNodeIndices GetNodeIndex(const UInt index) const
+        {
+            const UInt n = index / NumM();
+            const UInt m = index % NumM();
+            return {n, m};
```

Comment:
Should there be a check if index is out of bounds

Reply:
perhaps also a check on NumM not being zero, 
but maybe a check that index < NumM * NumN is sufficient